jetty-websocket/websocket-javax-common/src/main/java/org/eclipse/jetty/websocket/javax/common/JavaxWebSocketFrameHandler.java

```diff
@@ -596,10 +587,24 @@ private void acceptMessage(Frame frame, Callback callback)
 
     public void onPing(Frame frame, Callback callback)
     {
-        ByteBuffer payload = BufferUtil.copy(frame.getPayload());
-        coreSession.sendFrame(new Frame(OpCode.PONG).setPayload(payload), Callback.NOOP, false);
-        callback.succeeded();
-        coreSession.demand(1);
+        if (coreSession.isOutputOpen())
```

Comment:
Its not racy because we won't receive any new frames until we demand from the coreSession, and we don't demand until the write succeeds.

Reply:
In case of idle timeouts you have demand and the idle timeout fires and closes the output, as you send a Close frame but typically keep the input open to wait for the reply.

Also, if you received a message and give it to the application, the application could have spawned a thread so that you have demand again. At any moment the application can decide to close from the spawned thread.

Both not possible?